src/AWS.Deploy.CLI/Commands/GenerateDeploymentProjectCommand.cs

```diff
@@ -0,0 +1,185 @@
+            var saveDirectory = projectDirectoryFullPath + "DeploymentProject";
+
+            var suffixNumber = 0;
+            while (Directory.Exists(saveDirectory))
+                saveDirectory = projectDirectoryFullPath + $"DeploymentProject{++suffixNumber}";
+            
+            return saveDirectory;
+        }
+
+        private bool ValidateSaveDirectory(string saveDirectory)
+        {
+            var directoryCreatedByDeployTool = false;
+
+            if (Directory.Exists(saveDirectory))
+            {
+                CheckDirectoryEmpty(saveDirectory);
+            }
+            else
+            {
+                CreateNewDirectory(saveDirectory);
+                directoryCreatedByDeployTool = true;
+            }
+
+            CheckSaveDirectoryInsideProjectDirectory(saveDirectory, _session.ProjectDefinition, directoryCreatedByDeployTool);
+
+            return directoryCreatedByDeployTool;
+        }
+
+        private void CheckDirectoryEmpty(string saveDirectory)
+        {
+            if (Directory.GetFiles(saveDirectory).Length > 0 || Directory.GetDirectories(saveDirectory).Length > 0)
+            {
+                var errorMessage = "The directory specified for saving the CDK project is non-empty. " +
+                    "Please provide an empty directory path and try again.";
+                throw new InvalidSaveDirectoryForCdkProject(errorMessage);
+            }
+        }
+            
+
+        private void CreateNewDirectory(string saveDirectory)
+        {
+            try
+            {
+                Directory.CreateDirectory(saveDirectory);
+            }
+            catch (Exception e)
+            {
+                var errorMessage = "Failed to create a directory at the specified path as it contains one or more invalid characters";
+                throw new InvalidSaveDirectoryForCdkProject(errorMessage, e);
+            }
+        }
+
+        private void CheckSaveDirectoryInsideProjectDirectory(string saveDirectory,
+            ProjectDefinition projectDefinition, bool directoryCreatedByDeployTool)
+        {
+            var saveDirectoryFullPath = new DirectoryInfo(saveDirectory).FullName;
+            var projectDirectoryFullPath = new DirectoryInfo(projectDefinition.ProjectPath).Parent.FullName;
+
+            if (saveDirectoryFullPath.Contains(projectDirectoryFullPath + "\\", StringComparison.InvariantCultureIgnoreCase))
+            {
+                if (directoryCreatedByDeployTool)
+                    Directory.Delete(saveDirectory);
+
+                var errorMessage = "The directory used to save the CDK deployment project is contained inside of " +
+                    "the target deployment project directory. Please specify a different directory and try again.";
+                throw new InvalidSaveDirectoryForCdkProject(errorMessage);
+            }
+        }
+
+        private async Task<TryRunResult> CheckDirectoryUnderSourceControl(string saveDirectory)
```

Comment:
nit: you aren't using result in the form it is returned. `IsDirectoryUnderSourceControl` fits better which returns a boolean.

How are we going to handle if customer is using a different source control system like SVN?

Reply:
Changed the method name to `IsDirectoryUnderSourceControl` that returns a `bool`.